
syntax = "proto3";
import "google/protobuf/empty.proto";

package PythonIntelligenceAgent;

service PIAMessageService {
    rpc Publish (PublishMessage) returns (google.protobuf.Empty) {}
    rpc Subscribe (SubscribeMessage) returns (ResponseSubscribe) {}
    rpc UnSubscribe (SubscribeMessage) returns (ResponseSubscribe) {}
    rpc Request (RequestMessage) returns (ResponseMessage) {}
}

message PublishMessage {
    int32 message_id = 1;
    string publisher = 2;
    string receiver = 3;
    string publish_key = 4;
    string publish_value = 5;
}

message SubscribeMessage {
  int32 message_id = 1;
  string subscriber = 2;
  string receiver = 3;
  string subscribe_pattern = 4;
}

enum ResponseStatus {
    ALLOW = 0;
    DENIED = 1;
}

message ResponseSubscribe {
  int32 message_id = 1;
  ResponseStatus response_status = 2;
  SubscribeMessage subscribe_message = 3;
}

message RequestMessage {
  int32 message_id = 1;
  string requester = 2;
  string receiver = 3;
  string request_key = 4;
  string request_value = 5;
}

message ResponseMessage {
  int32 message_id = 1;
  string responser = 2;
  string receiver = 3;
  string response_key = 4;
  string response_value = 5;
  RequestMessage request_message = 6;
}